tooling/nargo/src/errors.rs

```diff
@@ -218,3 +218,29 @@ pub fn try_to_diagnose_runtime_error(
             .with_call_stack(source_locations),
     )
 }
+
+pub fn map_execution_error<F: AcirField>(
+    error: OpcodeResolutionError<F>,
+    call_stack: &[ResolvedOpcodeLocation],
+) -> ExecutionError<F> {
+    let call_stack = match &error {
+        OpcodeResolutionError::UnsatisfiedConstrain { .. }
+        | OpcodeResolutionError::IndexOutOfBounds { .. }
+        | OpcodeResolutionError::BrilligFunctionFailed { .. } => Some(call_stack.to_vec()),
+        _ => None,
+    };
+
+    let assertion_payload: Option<ResolvedAssertionPayload<F>> = match &error {
+        OpcodeResolutionError::BrilligFunctionFailed { payload, .. }
+        | OpcodeResolutionError::UnsatisfiedConstrain { payload, .. } => payload.clone(),
+        _ => None,
+    };
+
+    match assertion_payload {
+        Some(payload) => ExecutionError::AssertionFailed(
+            payload,
+            call_stack.expect("Should have call stack for an assertion failure"),
+        ),
+        None => ExecutionError::SolvingError(error, call_stack),
+    }
```

Comment:
I think this can be simplified to
```suggestion
    match &error {
        OpcodeResolutionError::BrilligFunctionFailed { payload: Some(payload), .. }
        | OpcodeResolutionError::UnsatisfiedConstrain { payload: Some(payload), .. } => {
            ExecutionError::AssertionFailed(payload.clone(), call_stack.to_vec())
        }
        _ => ExecutionError::SolvingError(error, Some(call_stack.to_vec())),
    }

```